enclave/core/tracee.c

```diff
@@ -0,0 +1,205 @@
+#include <openenclave/internal/report.h>
+#include <openenclave/internal/sgxtypes.h>
+#include <openenclave/internal/trace.h>
+#include <openenclave/internal/utils.h>
+
+static oe_log_filter_t g_active_log_filter = {0};
+static char g_logging_filename[MAX_FILENAME_LEN];
+static bool g_debug_allowed_enclave = false;
+
+const char* get_filename_from_path(const char* path)
+{
+    if (path)
+    {
+        for (size_t i = oe_strlen(path) - 1; i >= 0; i--)
+        {
+            if (path[i] == '/')
+            {
+                return &path[i + 1];
+            }
+        }
+    }
+    return path;
+}
+
+// Read an enclave's identity attribute to see to if it was signed as an debug
+// enclave
+bool is_enclave_debug_allowed()
+{
+    bool ret = false;
+#ifdef OE_USE_LIBSGX
+
+#if defined(__linux__)
+    oe_result_t result = 0;
+    td_t* td = oe_get_td();
+    uint8_t* report_buffer = NULL;
+
+    if (td->simulate)
+    {
+        // enclave in simulate mode is treated as debug_allowed
+        ret = true;
+    }
+    else
+    {
+        uint64_t report_buffer_size = OE_MAX_REPORT_SIZE;
+        const sgx_report_t* sgx_report = NULL;
+        oe_report_header_t* header = NULL;
+
+        report_buffer = (uint8_t*)oe_malloc(OE_MAX_REPORT_SIZE);
+        if (report_buffer == NULL)
+            goto done;
+
+        // get a report on the enclave itself for enclave identity information
+        report_buffer_size = OE_MAX_REPORT_SIZE;
+        result = oe_get_report(
+            0, NULL, 0, NULL, 0, report_buffer, &report_buffer_size);
+        if (result != OE_OK)
+            goto done;
+
+        header = (oe_report_header_t*)report_buffer;
+        sgx_report = (const sgx_report_t*)header->report;
+        ret = (sgx_report->body.attributes.flags & SGX_FLAGS_DEBUG) != 0;
+    }
+done:
+    if (report_buffer)
+        oe_free(report_buffer);
+#elif defined(_WIN32)
+    // WIN32 support is still under development.
+    // We will have to come back to handle this case
+    ret = true;
+#endif
+
+#else
+// When adding support for non-SGX solutions, we need to find a way to correctly
+// identify whether an enclave is debug-allowed.
+#endif
+    return ret;
+}
+
+/*
+**==============================================================================
+**
+** _handle_oelog_init()
+**
+** Handle the OE_ECALL_LOG_INIT from host and initialize SDK logging
+** configuration
+**
+**==============================================================================
+*/
+oe_result_t _handle_oelog_init(uint64_t arg)
+{
+    oe_result_t result = OE_FAILURE;
+    const char* filename = NULL;
+    oe_log_filter_t* filter = (oe_log_filter_t*)arg;
+
+    if (filter == NULL)
+    {
+        result = OE_INVALID_PARAMETER;
+        goto done;
+    }
+    /* Copy structure into enclave memory */
+    oe_memcpy(&g_active_log_filter, filter, sizeof(oe_log_filter_t));
+
+    if (!oe_is_outside_enclave((void*)filter, sizeof(oe_log_filter_t)))
```

Comment:
This function is still unsafe since it blindly copies over input to the global variable.

So the host can call init first with correct values and then when the logging system has been initialized, call init again with incorrect values and overwrite the global variable and tamper with the logging system.

Reply:
Great thank for spotting this subtle security issue. 
Addressed.

Reply:
Can you please explain why this solution is better than performing the copy between the two checks?

Reply:
Thanks for the reviewing, Istvan. 
I don;t see any difference between having this copy before those two checking or in between them in terms of the security concerns. 

Having this copying in-between those two check indeed is more efficient because it will the copy would have been avoided if the input arg was out-side-of-enclave, this is your comment trying to point out?

Reply:
What I was referring to was just moving `oe_memcpy(&g_active_log_filter, filter, sizeof(oe_log_filter_t));` to be between the two checks. And thus avoiding the need to perform a copy first onto the stack and then into global memory. Right now you are performing two copies and have this extra auxiliary variable, when there is an alternative solution to just move an already existing memcpy to a different place in the code.

There is no security impact, the code just looks confusing when it does not have to be.

Reply:
Got you point now. Thanks for the explanation.

anakrish pointed out that a malicious host might call _handle_oelog_init twice, which would cause the g_active_log_filter to get updated before both arg and path are validated. 
Having a local variable helps lock down both  values for eliminate such a security hole.

Reply:
anakrish is right that you should never dereference an untrusted pointer without first validating it. But this memcpy only dereferences the filter pointer and not filter.path (that one is dereferenced in the function called related to the path extraction), so it can be placed after the first validation, but before the second one.

The rule of thumb is:
- Any dereferencing of an untrusted pointer should be preceded by a range check corresponding to the range to be accessed.
- The first dereferencing of an untrusted pointer should be a memcpy into enclave memory (ideally volatile copy).
- Any subsequent dereferencing should be rewritten to use the enclave copy of the data instead.